File: app/src/sequential/java/app/we/go/emojidraw/features/practice/SequentialPracticePresenter.java
package app.we.go.emojidraw.features.practice;

import javax.inject.Inject;

import app.we.go.emojidraw.api.EmojiDetectionProvider;
import app.we.go.emojidraw.arch.di.qualifier.PracticeDuration;
import app.we.go.emojidraw.arch.di.scope.ActivityScope;
import app.we.go.emojidraw.data.EmojiToDrawProvider;


/**
 * This is al alternative implementation of {@link PracticeContract.Presenter}
 * for the <code>sequential</code> that is just meant to be used in the testing/development phase.
 * No point to get fancy with real decorators and stuff, plain old inheritance is OK.
 */
@ActivityScope
public class SequentialPracticePresenter extends PracticePresenter {

    private SequentialSharedPrefsHelper sharedPrefsHelper;


    @Inject
    public SequentialPracticePresenter(EmojiDetectionProvider detectionProvider,
                                       EmojiToDrawProvider emojiToDrawProvider,
                                       @PracticeDuration Integer duration,
                                       SequentialSharedPrefsHelper sharedPrefsHelper) {
        super(detectionProvider, emojiToDrawProvider, duration);
        this.sharedPrefsHelper = sharedPrefsHelper;
    }

    @Override
    protected boolean handleNextEmoji() {
        sharedPrefsHelper.increaseCurrentIndex();
        return super.handleNextEmoji();
    }


}

File: app/src/sequential/java/app/we/go/emojidraw/features/practice/SequentialPracticePresenter.kt
package app.we.go.emojidraw.features.practice

import app.we.go.emojidraw.api.EmojiDetectionProvider
import app.we.go.emojidraw.arch.di.ActivityScope
import app.we.go.emojidraw.arch.di.PracticeDuration
import app.we.go.emojidraw.data.EmojiToDrawProvider
import javax.inject.Inject


/**
 * This is al alternative implementation of [PracticeContract.Presenter]
 * for the `sequential` that is just meant to be used in the testing/development phase.
 * No point to get fancy with real decorators and stuff, plain old inheritance is OK.
 */
@ActivityScope
class SequentialPracticePresenter @Inject
constructor(detectionProvider: EmojiDetectionProvider,
            emojiToDrawProvider: EmojiToDrawProvider,
            @PracticeDuration duration: Int,
            private val sharedPrefsHelper: SequentialSharedPrefsHelper) :
    PracticePresenter(detectionProvider, emojiToDrawProvider, duration) {

    override fun handleNextEmoji(): Boolean {
        sharedPrefsHelper.increaseCurrentIndex()
        return super.handleNextEmoji()
    }


}
